Apply Ghost design to BlogPostCard and BlogPostCardListComponent

MoYeonWook opened the pull request:
# description
- ghost Casper 디자인을 BlogPostCard, BlogPostCardListComponent에 적용
- 기존 BlogPostHeader에서 사용하던 BlogPostHeaderTagItem을 BlogPostCard에 삽입
- BlogPostCard 디자인을 Large, Medium, Small로 분류 
- App.tsx에  main 태그를 추가

# commits

55fd58b Move BlogPostHeaderTagItem to shared directory
dc06a91 Replace PostPreview with PostHeaderData
69e249f Apply Ghost structure in BlogPostCard and BlogPostCardListComponent
aa1fa26 Set test-images to mock Data
91f61a2 Apply Ghost structure in BlogPostCard and BlogPostCardList
9c18776 Apply ghost css to BlogPostCard and BlogPostCardList
1d7e0f5 Insert main tag to App.tsx
a36de6b Change ClassName and Mention the Source of CSS


# 테스트사항
- home, searchpage에서 각각 의도한데로 postcard가 배치되는지 확인
- large, medium, small postcard가 homepage에서 구분되어 적용되는지 확인.
- 창 크기를 변화시키면서 이미지 크기 및  postcard의 불균형이 생기는지 확인

![image](https://github.com/WebBlogProject/web-blog-frontend/assets/87743224/e0368ef9-7fef-4b95-8ca7-5edac43fed4e)
![image](https://github.com/WebBlogProject/web-blog-frontend/assets/87743224/5c2e0ffb-d6d3-43c7-ae94-73c834e4e1d2)

MoYeonWook committed:
Move BlogPostHeaderTagItem to shared directory

related: #30

MoYeonWook committed:
Replace PostPreview with PostHeaderData

Initially, we separated PostPreview and PostHeaderData for components which not
need tag data. (such as BlogPostCard)
but now we have decided to use tags in those components. so PostPreview
is not needed.

related: #30

MoYeonWook committed:
Apply Ghost structure in BlogPostCard and BlogPostCardListComponent

Add or modify HTML tags based on the Ghost design and Add Tag
component to BlogPostCard

related: #30

MoYeonWook committed:
Set test-images to mock Data

related: #30

MoYeonWook committed:
Apply Ghost structure in BlogPostCard and BlogPostCardList

Add or modify HTML tags based on the Ghost design and Add Tag
component to BlogPostCard

related: #30

MoYeonWook committed:
Apply ghost css to BlogPostCard and BlogPostCardList

related: #30

MoYeonWook committed:
Insert main tag to App.tsx

The main tag is needed to add basic padding

related: #30

MoYeonWook committed:
Change ClassName and Mention the Source of CSS

Changed the classname according to project's convention.

related: #30

SonDongGyu commented on src/presentation/shared/components/css/BlogPostCard.css:
```diff
@@ -1,35 +1,60 @@
+* Copy or variation of .gh-canvas > * style of screen.css.
+* Original file: https://github.com/TryGhost/Casper/blob/main/assets/css/screen.css#L866
+*/
+.post-card {
```
naming 규칙을 통일하면 좋을 것 같습니다. `.PostCard` 로 하면 그 하위의 요소들이 현재 `PostCard-...` 의 형태이므로 일관성이 있어 더 좋을 것 같습니다.

MoYeonWook replied:
동의합니다

SonDongGyu commented on src/App.css:
```diff
@@ -32,6 +32,15 @@
+* Copy or variation of .gh-canvas > * style of screen.css.
+* Original file: https://github.com/TryGhost/Casper/blob/main/assets/css/screen.css#L77
+*/
+.site-main {
```
`.BlogMain`? 일관성이 유지되면 좋을 것 같아요

SonDongGyu commented on src/presentation/shared/components/ts/BlogPostCardListComponent.tsx:
```diff
@@ -26,17 +26,25 @@ function BlogPostCardListComponent({
               thumbnailUrl={post.thumbnailUrl}
               key={post.id}
+              tagList={post.tagList}
+              cardSize={
```
cardSize를 외부에서 주입하는 형식은 좋은 것 같습니다. 조금 더 개선해볼 수 있을 것 같아요.
현재 개선될만한 포인트는 아래와 같습니다.
1. 현재 조건들이 실질적으로는 Home이나 Search와 같이 이 component를 사용하는 외부에 의해 결정되는데 이걸 내부에서 판단하고 있음 -> 이 경우 page가 많아지고 그에 따라 design이 다양해지면 code가 굉장히 복잡해질 여지가 있음
2. Large, Medium, Small은 일종의 고정된 type이므로 humman error를 최소화하기 위해 상수화하면 좋을 것 같음
3. 실질적으로 style에 영향을 주지 않는 HomeCardLayout과 같은 className이 size를 결정하기 위해 남아있음 -> 이는 추후에 page와 design이 더 다양해지면 불필요한 className들을 계속 추가해야하는 단점이 있음

이에 따라 아래와 같이 변경하는 것을 제안합니다.
(참고, 아래의 제안은 pseudocode 임을 유의 부탁드립니다,)

1. card size에 대한 타입 정의
```
enum BlogCardType {
    LARGE_CARD,
    MEDIUM_CARD,
    SMALL_CARD
}
```

2. card size를 실제 size pattern을 결정하는 외부에서 함수의 형태로 전달
```
type BlogPostCardListComponentProps = {
  ...,
  cardSizeGetter: (number) -> BlogCardType;
};
```
예를 들어 home에서 주입한다면
```
fun getCardTypeByIndex(index: number) {
  switch (index) {
      case 0:
           return BlogCardType.LARGE_CARD;
       case 1:
       case 2:
            return BlogCardType.MEDIUM_CARD;
       default:
             return BlogCardType.SMALL_CARD;
  }
}
```
위와 같은 함수를 `BlogPostCardListComponent`의 prop으로 넘겨줌

3. 불필요한 HomeCardLayout 같은 className 주입 제거

MoYeonWook replied:
함수를 주입시켜서 용도 별로 다른 컴포넌트를 구현할 생각은 못해봤는데 좋은 방법인 것 같습니다

SonDongGyu reviewed (commented):
Commit dc06a9188075cb176d76d2126d25fd07c0828c48 의 message에 typo 있습니다.
`we seperated` -> `we separated`

MoYeonWook committed:
Add CardTypeGetter props in BlogPostCardListComponent

CardType Selector function was declared inside the BlogPostCardListComponent.
Now It is injected from parent component.
This architectural change helps to remove unneeded ClassName (such as
HomeCardLayout).

MoYeonWook committed:
Apply Reviewer's Comment

MoYeonWook commented:
> Commit [dc06a91](https://github.com/WebBlogProject/web-blog-frontend/commit/dc06a9188075cb176d76d2126d25fd07c0828c48) 의 message에 typo 있습니다. `we seperated` -> `we separated`

해당 작업 수행을 위해 rebase를 하려고 했지만 이미 우진님 커밋들과 머지되어서 rebase 결과물이 이상하게 나오더라구요. 추후 관련 작업이 가능한지 질문드리고 싶네요

SonDongGyu commented on src/application/types/BlogCardType.ts:
```diff
@@ -0,0 +1,7 @@
+enum BlogCardType {
```
https://github.com/WebBlogProject/web-blog-frontend/pull/41/commits/74ce704363a964011776aea9538001a37cdfcc4e

이전에 작업했던 commit message 참고해보면 typescript에서 enum을 바로 쓰는걸 그렇게 지향하는것 같지는 않더라고요. 한 번 검토 부탁드립니다. (제 생각엔 저희 정도 규모에서는 크게 상관 없을 것 같아서 blocker는 아닙니다)

SonDongGyu reviewed (approved):
고생하셨습니다.
추후에 search result 쪽이나 header 쪽 margin 정도 한 번 통일해보면 좋을 것 같기도 하네요.

SonDongGyu commented:
> > Commit [dc06a91](https://github.com/WebBlogProject/web-blog-frontend/commit/dc06a9188075cb176d76d2126d25fd07c0828c48) 의 message에 typo 있습니다. `we seperated` -> `we separated`
> 
> 해당 작업 수행을 위해 rebase를 하려고 했지만 이미 우진님 커밋들과 머지되어서 rebase 결과물이 이상하게 나오더라구요. 추후 관련 작업이 가능한지 질문드리고 싶네요

아마 중간에 merge commit이 들어간 상태라 그렇지 않을까 싶네요, 중간에 merge commit 있으면 rebase 결과 이상하게 보통 나오는 걸로 알고있어서요. 보통 저는 그럴 때 그냥 reset --hard로 돌아가서 cherry-pick 하나씩 수작업을 합니다. ~머리로 안 되면 몸이 고생하는 거죠~

Woojinger commented on src/presentation/shared/components/ts/BlogPostCard.tsx:
```diff
@@ -28,7 +32,7 @@ function BlogPostCard({
 
   return (
-    <article className="Post-card" onClick={onClickPostCard}>
+    <article className={'post-card ' + cardSize} onClick={onClickPostCard}>
```
className을 동적으로 설정하는게 저는 조금 위험하다고 생각하는데요.

return 문 앞에 switch case문 같은 것을 넣어서 className이 정의되지 않은 이름이 되는 것을 방지하는 것은 어떨까요?

MoYeonWook replied:
cardType을 정해주는 함수(cardTypeGetter)를 외부에서 주입 받고 있기는 하지만 cardsize가 BlogCardType enum으로 한정되어 있어서 다른 예외의 string이 들어올 수 없습니다. 제한된 범위 내에서 cardsize가 결정되므로 예외처리는 불필요하다고 생각합니다.

Woojinger replied:
이전 변경에 코멘트 단 것이라 enum임을 확인 못했습니다.
이상 없겠네요. 감사합니다!

Woojinger commented on src/presentation/home/pages/ts/BlogHomePage.tsx:
```diff
@@ -25,6 +26,18 @@ function BlogHomePage() {
     homeResult.nextPage
   );
 
+  const getCardTypebyIndex = useCallback((index: number) => {
```
```getCardTypebyIndex``` -> ```getCardTypeByIndex```가 어떨까요?

MoYeonWook replied:
동의합니다

Woojinger commented on src/presentation/searchPage/pages/ts/BlogSearchResultPage.tsx:
```diff
@@ -47,6 +48,10 @@ function BlogSearchResultPage() {
     }
   }, [stateQuery, query, dispatch, resetIntersectingState]);
 
+  const getCardTypebyIndex = useCallback(() => {
```
```getCardTypebyIndex```가 index를 받지 않는 함수인데도 에러가 안나는군요!
정말 놀라운 타입스크립트의 세계...

MoYeonWook replied:
네 저도 그 부분이 신기하긴 했습니다.  인덱스 파라미터가 안 쓰여서 그런 것 같기도 하고..

Woojinger reviewed (commented):
리뷰 늦어서 죄송합니다.
코멘트 확인 한번 부탁드립니다!

Woojinger reviewed (approved):
> getCardTypebyIndex -> getCardTypeByIndex가 어떨까요?

이 부분만 수정해주시고 머지하시면 될 것 같습니다
수고하셨습니다
